x-pack/plugins/reporting/server/browsers/install.ts

```diff
@@ -19,37 +23,55 @@ const chmod = promisify(fs.chmod);
 interface Package {
   platforms: string[];
 }
-interface PathResponse {
-  binaryPath: string;
-}
 
 /**
  * "install" a browser by type into installs path by extracting the downloaded
  * archive. If there is an error extracting the archive an `ExtractError` is thrown
  */
-export async function installBrowser(
-  logger: LevelLogger,
-  browser: BrowserDownload,
-  installsPath: string
-): Promise<PathResponse> {
-  const pkg = browser.paths.packages.find((p: Package) => p.platforms.includes(process.platform));
-
-  if (!pkg) {
-    throw new Error(`Unsupported platform: ${JSON.stringify(browser, null, 2)}`);
+export class BrowserInstaller {
+  public readonly binaryPath$: Rx.Subject<string>;
+
+  constructor(browser: BrowserDownload, config: ReportingConfig, logger: LevelLogger) {
+    this.binaryPath$ = new Rx.Subject<string>();
+    this.backgroundInstall(browser, config, logger);
   }
 
-  const binaryPath = path.join(installsPath, pkg.binaryRelativePath);
-  const binaryChecksum = await md5(binaryPath).catch(() => '');
+  async backgroundInstall(browser: BrowserDownload, config: ReportingConfig, logger: LevelLogger) {
+    const captureConfig = config.get('capture');
+    const { autoDownload, type: browserType } = captureConfig.browser;
+    if (autoDownload) {
+      await ensureBrowserDownloaded(browserType, logger);
+    }
 
-  if (binaryChecksum !== pkg.binaryChecksum) {
-    const archive = path.join(browser.paths.archivesPath, pkg.archiveFilename);
-    logger.debug(`Extracting [${archive}] to [${binaryPath}]`);
-    await extract(archive, installsPath);
-    await chmod(binaryPath, '755');
-  }
+    const pkg = browser.paths.packages.find((p: Package) => p.platforms.includes(process.platform));
+    if (!pkg) {
+      throw new Error(`Unsupported platform: ${JSON.stringify(browser, null, 2)}`);
+    }
+
+    const dataDir = await config.kbnConfig.get('path', 'data').pipe(first()).toPromise(); // FIXME: plain subscribe
+    const binaryPath = path.join(dataDir, pkg.binaryRelativePath);
+    this.binaryPath$.next(binaryPath);
 
-  logger.debug(`Browser installed at ${binaryPath}`);
-  return {
-    binaryPath,
-  };
+    try {
+      const binaryChecksum = await md5(binaryPath).catch(() => '');
+
+      if (binaryChecksum !== pkg.binaryChecksum) {
+        const archive = path.join(browser.paths.archivesPath, pkg.archiveFilename);
+        logger.info(`Extracting [${archive}] to [${binaryPath}]`);
+        await extract(archive, dataDir);
+        await chmod(binaryPath, '755');
+      }
+    } catch (error) {
+      if (error.cause && ['EACCES', 'EEXIST'].includes(error.cause.code)) {
+        logger.error(
+          `Error code ${error.cause.code}: Insufficient permissions for extracting the browser archive. ` +
+            `Make sure the Kibana data directory (path.data) is owned by the same user that is running Kibana.`
+        );
+      }
+
+      throw error; // reject the promise with the original error
```

Comment:
TODO: test promise rejection

Reply:
Done now since we have a plugin.ts test.